### ISSUE TITLE
Invalid block returned by buttons

### ISSUE THREAD

**Issue by LemADEC** [kyeiti/UndergroundBiomesConstructs#40]
As of 1.12-1.2.1, exceptions are reported by EnderIO 1.12.2-5.0.37:
```
[20:41:51] [main/ERROR] [enderio]: ItemBlock exterminatorjeff.undergroundbiomes.common.itemblock.ButtonItemBlock@3912ce25 returned an unregistered block from getBlock(). This is a major bug in the mod 'Underground Biomes'.
[20:41:51] [main/ERROR] [enderio]: ItemBlock exterminatorjeff.undergroundbiomes.common.itemblock.ButtonItemBlock@45acbdb1 returned an unregistered block from getBlock(). This is a major bug in the mod 'Underground Biomes'.
[20:41:51] [main/ERROR] [enderio]: ItemBlock exterminatorjeff.undergroundbiomes.common.itemblock.ButtonItemBlock@288e464e returned an unregistered block from getBlock(). This is a major bug in the mod 'Underground Biomes'.
[20:41:51] [main/ERROR] [enderio]: ItemBlock exterminatorjeff.undergroundbiomes.common.itemblock.ButtonItemBlock@3790b17e returned an unregistered block from getBlock(). This is a major bug in the mod 'Underground Biomes'.
[20:41:51] [main/ERROR] [enderio]: ItemBlock exterminatorjeff.undergroundbiomes.common.itemblock.ButtonItemBlock@7cbe8b21 returned an unregistered block from getBlock(). This is a major bug in the mod 'Underground Biomes'.
```

**Comment by kreezxil** [comment]
does this actually affect the world?, can I get a full log?

**Comment by LemADEC** [comment]
Refer to Aang23 fork which is way more updated and already discuss that issue, see above link

**Comment by Aang23** [comment]
@kreezxil No, this doesn't affect gameplay or the world at all, this isn't even causing any bugs, but this may in the future or with other mods (EnderIO as an example), so this will need to be fixed.
No direct link with that issue, but I saw that you forked UBC too, so I suppose you were planning to work on that mod?

**Comment by kreezxil** [comment]
@aang23 I see you have a pull request, while I have control of the project
now on CurseForge, I don't of this. I did make a fork. I pulled you changes
in and encountered a massive memory race condition. I then looked at the
mod without your changes and the memory race condition is still there.
Wondering if you encountered on your end.

Also,  your changes seem to indicate you are way more adept with a mod of
this complexity than I am. Would you be interested in taking it over?

I can add you to the CurseForge project as an equal to myself and then
point the issues and source to your fork.

**Comment by LemADEC** [comment]
What did you see that qualify as a 'Memory race condition'? Any details?

**Comment by Aang23** [comment]
@kreezxil For sure. I actually noticed that as soon as I started working on that mod. I was planning to rewrite some core parts of the mods, mainly the block system and the world gen things. This race condition things seems to be less impactful on some forge versions. The village bug actually comes from that too.

Yes, I would be interested in taking over, and that would be better to do that one the original mod page. Thanks for that proposition ! I'm gonna be moving all the builds that are on my fork's page to the "real" one, to keep that clean. I will also be giving you write access to my fork's git, so that if need / want to do something, you can.

**Comment by kreezxil** [comment]
@LemADEC 
> What did you see that qualify as a 'Memory race condition'? Any details?

install the mod by itself, run it, it will eat all the allocated ram and crash with a stack overflow in buffer memory error.

@Aang23 
> @kreezxil For sure. I actually noticed that as soon as I started working on that mod. I was planning to rewrite some core parts of the mods, mainly the block system and the world gen things. This race condition things seems to be less impactful on some forge versions. The village bug actually comes from that too.
> 
> Yes, I would be interested in taking over, and that would be better to do that one the original mod page. Thanks for that proposition ! I'm gonna be moving all the builds that are on my fork's page to the "real" one, to keep that clean. I will also be giving you write access to my fork's git, so that if need / want to do something, you can.

I don't have access to the main repository which is actually zeno's nor kyeti's but you can see on CurseForge that it was transferred to me recently. What is your curseforge handle?

**Comment by LemADEC** [comment]
@kreezxil I've already made 3 different 1.12.2 servers running with UBC and didn't see a memory or stack error. I could use it in my mod's dev space without such concern.
I did fix quite a few bugs on the 1.7.10 Geno branch, so maybe some of those aren't fixed in the port?

**Comment by kreezxil** [comment]
I've noticed it with forge builds 2768 and higher. Try allocating 1 to 2gb
to the mod and watch. don't put the mod in a pack where you are then gonna
allocate 8 to 16gb of ram to compensate for the 100's of other mods you
added.

**Comment by Aang23** [comment]
@LemADEC What bugs did you fix ?

**Comment by LemADEC** [comment]
Asiekierka fixed a world generation lag (notifications were enabled during block placement).
I've:
- fixed a crash with missing tile entities for button/stairs/etc. (n/a for 1.12.2 version)
- added monster egg stone variations
- fixed missing ore registration on ores and slabs
- fixed various display names and harvest levels issues with ores
- fixed a hashcode issue in internal collections
- fixed invalid slab variations in villages
- fixed the command error handling
- fixed dimension blacklist so it actually works
- fixed biome decoration error handling to narrow down an issue in another mod (CoG probably)

I remember there was a lag issue with the mining speed modifier event, but I never really fixed it.

**Comment by Aang23** [comment]
I'm gonna port some of those fixes, even if I have already done some of them. I can confirm that some were not ported. Note that some aren't needed currently in 1.12.2.

**Comment by LemADEC** [comment]
There was also a lot of new mods support added.
Here are some I've readded on 1.12.2 for my modpack:
[ores.zip](https://github.com/kyeiti/UndergroundBiomesConstructs/files/2727554/ores.zip)

**Comment by Aang23** [comment]
Thanks, I'll add them in the next release.

**Comment by LemADEC** [comment]
I see autoreglib is in the dev space libs folder for Aang23 fork. There's a pretty bad memory leak in that library that was fixed recently. I suggest to add just an API instead of the full mod.

I've tested a modpack with just UBC, forge 2796, 1GB RAM allocated. I didn't see a crash and it didn't use the full RAM allocated. Retrying without UBC (i.e. Vanilla forge), I didn't see much of a difference in memory usage.

**Comment by Aang23** [comment]
Autoreglib is never used, and not even required unless Quark is installed, that was only needed to add support for Speleothems. I'll look if adding an API only is possible.

There's no crashes for me too, but the RAM usage goes up quickly...

**Comment by LemADEC** [comment]
As long as the jar is in the libs folder, it'll be running in your debug space. This can be quickly problematic when testing the mod.
What kind of interfaces are used in Quark for UBC?

**Comment by Aang23** [comment]
You should be able to see that there : https://github.com/Aang23/UndergroundBiomesConstructs/blob/master/src/main/java/exterminatorjeff/undergroundbiomes/common/block/IgneousSpeleothem.java

That problematic for sure, I think that I'm gonna be switching to the API soon.
